### PR TITLE
Programmes 6763 twitter fb metadata

### PR DESCRIPTION
**Pull request by cisco87**
https://jira.dev.bbc.co.uk/browse/PROGRAMMES-6763

Gallery Item V3 Tags: http://sandbox.bbc.co.uk/programmes/p06t550j/p06t54j0
```
twitter:card = gallery 
twitter:title = BBC One - Doctor Who, New Year's Day Special Gallery 
twitter:image = https://ichef.bbci.co.uk/images/ic/1200x675/p06t551w.jpg 
twitter:description = New images released from the Doctor Who New Year's Day episode. 
twitter:image:alt = New images released from the Doctor Who New Year's Day episode.
twitter:url = http://sandbox.bbc.co.uk/programmes/p06t550j 

og:site_name = BBC
og:title = BBC One - Doctor Who, New Year's Day Special Gallery
og:type = website
og:url = http://sandbox.bbc.co.uk/programmes/p06t550j/p06t54j0
og:description = New images released from the Doctor Who New Year's Day episode. 
og:image = https://ichef.bbci.co.uk/images/ic/1200x675/p06t551w.jpg

fb:admins = 100004154058350
fb:pages = 6025943146,7397061762,7519460786,7833211321,8244244903,8251776107,8585725981,21750735380,80758950658,125309456546,130593816777,154344434967,228735667216,260212261199,260967092113,294662213128,295830058648,304314573046,401538510458,107909022566650,118883634811868,129044383774217,156060587793370,156400551056385,163571453661989,168895963122035,185246968166196,193022337414607,193435954068976,194575130577797,215504865453262,239931389545417,273726292719943,283348121682053,286567251709437,292291897588734,310719525611571,317278538359186,413132078795966,470911516262605,512423982152360,647687225371774,658551547588605,742734325867560,944295152308991,958681370814419,1143803202301544,1159932557403143,1392506827668140,1411916919051820,1477945425811579,1659215157653827,1731770190373618 
```

Gallery V3 Tags: http://sandbox.bbc.co.uk/programmes/p06t550j
```
<meta name="twitter:card" content="gallery" />
<meta name="twitter:image0" content="https://ichef.bbci.co.uk/images/ic/560xn/p06t54j0.jpg" />
<meta name="twitter:image1" content="https://ichef.bbci.co.uk/images/ic/560xn/p06t54nr.jpg" />
<meta name="twitter:image2" content="https://ichef.bbci.co.uk/images/ic/560xn/p06t54bq.jpg" />

twitter:title = BBC One - Doctor Who, New Year's Day Special Gallery 
twitter:image = https://ichef.bbci.co.uk/images/ic/1200x675/p06t551w.jpg 
twitter:description = New images released from the Doctor Who New Year's Day episode.
twitter:image:alt = New images released from the Doctor Who New Year's Day episode. 
twitter:url = http://sandbox.bbc.co.uk/programmes/p06t550j

og:site_name = BBC 
og:title = BBC One - Doctor Who, New Year's Day Special Gallery
og:type = website 
og:url = http://sandbox.bbc.co.uk/programmes/p06t550j/p06t54j0
og:description = New images released from the Doctor Who New Year's Day episode.
og:image = https://ichef.bbci.co.uk/images/ic/1200x675/p06t551w.jpg

fb:admins = 100004154058350
fb:pages = 6025943146,7397061762,7519460786,7833211321,8244244903,8251776107,8585725981,21750735380,80758950658,125309456546,130593816777,154344434967,228735667216,260212261199,260967092113,294662213128,295830058648,304314573046,401538510458,107909022566650,118883634811868,129044383774217,156060587793370,156400551056385,163571453661989,168895963122035,185246968166196,193022337414607,193435954068976,194575130577797,215504865453262,239931389545417,273726292719943,283348121682053,286567251709437,292291897588734,310719525611571,317278538359186,413132078795966,470911516262605,512423982152360,647687225371774,658551547588605,742734325867560,944295152308991,958681370814419,1143803202301544,1159932557403143,1392506827668140,1411916919051820,1477945425811579,1659215157653827,1731770190373618
```

Gallery V2 Tags: https://www.bbc.co.uk/programmes/p06t550j/p06t54j0
```
twitter:title = New Year's Day Special Gallery - Doctor Who - BBC One debugger
twitter:card = gallery debugger
twitter:image0 = https://ichef.bbci.co.uk/images/ic/560xn/p06t54j0.jpg
twitter:image1 = https://ichef.bbci.co.uk/images/ic/560xn/p06t54nr.jpg debugger
twitter:image2 = https://ichef.bbci.co.uk/images/ic/560xn/p06t54bq.jpg
twitter:description = New images released from the Doctor Who New Year's Day episode.
twitter:url = https://www.bbc.co.uk/programmes/p06t550j

og:site_name = BBC
og:title = New Year's Day Special Gallery - Doctor Who - BBC One
og:url = https://www.bbc.co.uk/programmes/p06t550j 
og:type = website
og:description = New images released from the Doctor Who New Year's Day episode. 
og:image = https://ichef.bbci.co.uk/images/ic/1200x675/p06t551w.jpg

fb:admins = 100004154058350
fb:pages = 194575130577797,7397061762,7833211321,129044383774217,401538510458,8251776107,273726292719943,107909022566650,21750735380,1477945425811579,944295152308991,154344434967,8244244903,512423982152360,156060587793370,215504865453262,260967092113,1392506827668140,193435954068976,7519460786,1143803202301544,470911516262605,413132078795966,118883634811868,130593816777,228735667216,80758950658,260212261199,294662213128,283348121682053,295830058648,239931389545417,304314573046,310719525611571,647687225371774,1159932557403143,286567251709437,1731770190373618,125309456546,742734325867560,168895963122035,292291897588734,658551547588605,156400551056385,185246968166196,1411916919051820,958681370814419,317278538359186,8585725981,163571453661989,6025943146,193022337414607,1659215157653827
```

Gallery Item V2 Tags: https://www.bbc.co.uk/programmes/p06t550j/p06t54j0
```
twitter:title = New Year's Day Special Gallery - Doctor Who - BBC One
twitter:card = gallery 
twitter:image0 = https://ichef.bbci.co.uk/images/ic/560xn/p06t54j0.jpg
twitter:image1 = https://ichef.bbci.co.uk/images/ic/560xn/p06t54nr.jpg
twitter:image2 = https://ichef.bbci.co.uk/images/ic/560xn/p06t54bq.jpg
twitter:description = New images released from the Doctor Who New Year's Day episode.
twitter:url = https://www.bbc.co.uk/programmes/p06t550j

og:site_name = BBC
og:title = New Year's Day Special Gallery - Doctor Who - BBC One
og:url = https://www.bbc.co.uk/programmes/p06t550j
og:type = website
null = #F9F8F6
og:description = New images released from the Doctor Who New Year's Day episode.
og:image = https://ichef.bbci.co.uk/images/ic/1200x675/p06t551w.jpg

fb:admins = 100004154058350 debugger eval code:1:77
fb:pages = 194575130577797,7397061762,7833211321,129044383774217,401538510458,8251776107,273726292719943,107909022566650,21750735380,1477945425811579,944295152308991,154344434967,8244244903,512423982152360,156060587793370,215504865453262,260967092113,1392506827668140,193435954068976,7519460786,1143803202301544,470911516262605,413132078795966,118883634811868,130593816777,228735667216,80758950658,260212261199,294662213128,283348121682053,295830058648,239931389545417,304314573046,310719525611571,647687225371774,1159932557403143,286567251709437,1731770190373618,125309456546,742734325867560,168895963122035,292291897588734,658551547588605,156400551056385,185246968166196,1411916919051820,958681370814419,317278538359186,8585725981,163571453661989,6025943146,193022337414607,1659215157653827
```

### CODE REVIEW

**Comment by programmes-developers** [comment]
Can one of the admins verify this patch?

File: src/Controller/Profiles/ShowController.php [hunk 1]
```diff
@@ -68,8 +68,14 @@ public function __invoke(string $key, string $slug, Request $request, ProfileSer
          * @var IsiteResult[] $response
          */
         $response = $isiteService->setChildrenOn([$profile], $profile->getProjectSpace(), $this->getPage())->wait(true);
-        $extraProfiles = array_merge($profile->getChildren(), $profile->getParents()); // We want to fetch the children of the main profiles children and parents.
-        $grandChildPromise = $isiteService->setChildrenOn($extraProfiles, $profile->getProjectSpace());
+        $profilesToFindChildrenFor = [];
```
Comment (cisco87) [inline src/Controller/Profiles/ShowController.php]: Mhhh I have no idea how it ended up here ;o

File: src/Controller/BaseController.php [hunk 2]
```diff
@@ -264,18 +282,19 @@ protected function renderWithoutChrome(string $view, array $parameters = [])
...
+     * @param string $url           The URL to redirect to
+     * @param int    $status        The status code to use for the Response
      * @param int    $cacheLifetime Seconds the response should be cached for
      *
      * @return RedirectResponse
      */
     protected function cachedRedirect($url, $status = 302, int $cacheLifetime = null): RedirectResponse
     {
-        if ($cacheLifetime !== null) {
+        if (!is_null($cacheLifetime)) {
```
Comment (Bobkidbob) [inline src/Controller/BaseController.php]: Why have you unoptimised this?

File: src/Controller/BaseController.php [hunk 3]
```diff
@@ -225,37 +227,53 @@ protected function renderWithChrome(string $view, array $parameters = [])
...
+            $this->context,
+            $cosmosInfo,
+            $istatsAnalyticsLabelsInstance,
+            $this->getCanonicalUrl() . $urlQueryString
+        );
 
         $parameters = array_merge([
             'orb' => $orb,
-            'meta_context' => new MetaContext($this->context, $this->getCanonicalUrl(), $this->getMetaNoIndex(), $this->overridenDescription),
-            'comscore' => (new ComscoreAnalyticsLabels($this->context, $cosmosInfo, $istatsAnalyticsLabelsInstance, $this->getCanonicalUrl() . $urlQueryString))->getComscore(),
```
Comment (Bobkidbob) [inline src/Controller/BaseController.php]: I would keep this inline unless it's used more than once.

File: src/Ds2013/PresenterFactory.php [hunk 4]
```diff
@@ -344,6 +344,8 @@ public function contentBlockPresenter(
         ));
     }
 
+
+
```
Comment (Bobkidbob) [inline src/Ds2013/PresenterFactory.php]: Unnecessary whitespace.

File: src/DsShared/PresenterFactory.php [hunk 5]
```diff
@@ -23,6 +23,7 @@ public function entityContextPresenter(
         return new EntityContextPresenter($context, $options);
     }
 
+
```
Comment (Bobkidbob) [inline src/DsShared/PresenterFactory.php]: Unnecessary whitespace.

File: src/Controller/FindByPid/GalleryController.php [hunk 6]
```diff
@@ -41,7 +42,8 @@ public function __invoke(
             'network' => $network,
             'galleries' => $galleries,
             'brand' => $brand,
-        ]);
+            'isGallery' => $isGallery,
+         ]);
```
Comment (Bobkidbob) [inline src/Controller/FindByPid/GalleryController.php]: Think there's an extra space here
```suggestion
        ]);
```

File: templates/find_by_pid/gallery.html.twig [hunk 7]
```diff
@@ -2,6 +2,24 @@
 
 {% block title %}{{ meta_context.titlePrefix() }}{% endblock %}
 
+{% block twitter_tags %}
+    {% if isGallery %}
```
Comment (Bobkidbob) [inline templates/find_by_pid/gallery.html.twig]: Is this clear enough? Surely all pages using the gallery presenter are galleries.

File: src/Controller/FindByPid/GalleryController.php [hunk 8]
```diff
@@ -31,6 +31,7 @@ public function __invoke(
         $brand = $programme->getTleo();
         $network = $programme->getMasterBrand()->getNetwork()->getName();
         $galleries = $programmesAggregationService->findDescendantGalleries($brand, $siblingLimit);
+        $hasImageHighlighted = !empty($imagePid);
```
Comment (BjornTwachtmann) [inline src/Controller/FindByPid/GalleryController.php]: isn't this basically the same as !showFirstImage? Do we need two variants of that boolean?

Comment (cisco87) [reply src/Controller/FindByPid/GalleryController.php]: Wouldn't it be breaking the SRP by using the same variable calculated the same way but serving 2 different use cases? I think it would cause stretching in the future. I thought that was used to decide whether the page had to autoload the first image in the gallery, mine serves to understand whether it's serving any image in the gallery that might change... Or not

Comment (BjornTwachtmann) [reply src/Controller/FindByPid/GalleryController.php]: Personally I would just name it better, "isListingPage" or "isItemPage". That's really what it's working out, they're not separate things, just confusingly named.

File: templates/find_by_pid/gallery.html.twig [hunk 9]
```diff
@@ -2,6 +2,27 @@
...
 
+{% block twitter_tags %}
+    {% if hasImageHighlighted %}
+        <meta name="twitter:card" content="photo" />
+        <meta name="twitter:description" content="{{ image.getShortSynopsis() }}" />
+        <meta name="twitter:image" content="{{ image.getUrl(560) }}">
+        <meta name="twitter:image:alt" content="{{ image.getShortSynopsis() }}" />
+    {% else %}
+        <meta name="twitter:card" content="gallery" />
+        <meta name="twitter:description" content="{{ gallery.getSynopses().getShortSynopsis() }}" />
```
[the remark this answers is not in the document]
Comment (cisco87) [reply templates/find_by_pid/gallery.html.twig]: Are you sure? I think I found a getSynopsis in V2 which was returning 
```
    public function getSynopsis()
   {
       $synopses = $this->group->getSynopses();
       return $synopses['shortest'];
   }
```

File: templates/find_by_pid/gallery.html.twig [hunk 10]
```diff
@@ -2,6 +2,27 @@
...
+        {% for index, image in images|slice(0, 4) %}
+            <meta name="twitter:image{{ index }}" content="{{ image.getUrl(560) }}" />
+        {% endfor %}
+    {% endif %}
+
+    {% if meta_context.canonicalUrl() %}
```
[the remark this answers is not in the document]
Comment (cisco87) [reply templates/find_by_pid/gallery.html.twig]: If you want I can remove it but in base_2013 template it's 

```
        {% if meta_context.canonicalUrl() %}<meta property="og:url" content="{{ meta_context.canonicalUrl() }}">{% endif %}
```

I thought it was put there for a reason and didn't want to make assumptions

File: templates/find_by_pid/gallery.html.twig (hunk 9 above)
[the remark this answers is not in the document]
Comment (BjornTwachtmann) [reply templates/find_by_pid/gallery.html.twig]: Yeah, it wasn't calling that method though (where is that anyway)? I sent you the links in slack, this is what it's calling https://github.com/bbc/programmes/blob/develop/webapp/php/lib/BBC/Programmes/View/Gallery.php#L69

File: src/Controller/FindByPid/GalleryController.php (hunk 8 above)
Comment (cisco87) [reply src/Controller/FindByPid/GalleryController.php]: Ok then I will change the occurrences of `showFirstImage` to something more generic, like `hasImageHighlighted`

**Comment by Bobkidbob** [comment]
Test this please.
